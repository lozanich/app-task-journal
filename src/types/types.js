export const types = {
  login: "[AUTH] loggeado en la aplicacion",
  logout: "[AUTH] NO loggeado en la aplicacion",

  uiSetError: "[UI] Set Error",
  uiRemoveError: "[UI] Remove Error",

  uiStartLoading: "[UI] Start loading",
  uiEndLoading: "[UI] End loading",

  notesAddNew: "[NOTES] New note",
  notesActive: "[NOTES] Set active note",
  notesLoad: "[NOTES] Load notes",
  notesUpdated: "[NOTES] Update note",
  notesFileUrl: "[NOTES] Update image url",
  notesDelete: "[NOTES] Delete note",
  notesLogoutCleanning: "[NOTES] Logout cleaning notes",
};
