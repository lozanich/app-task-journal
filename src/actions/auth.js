import Swal from "sweetalert2";
import { firebase, googleAuthProvider } from "../firebase/firebaseConfig";
import { types } from "../types/types";
import { endLoadingAction, startLoadingAction } from "./ui";

export const startLoginEmailPassword = (email, password) => {
  return (dispatch) => {
    // setTimeout(() => {
    //   dispatch(login(987456, "Lozano"));
    // }, 3500);
    dispatch(startLoadingAction());
    firebase
      .auth()
      .signInWithEmailAndPassword(email, password)
      .then(({ user }) => {
        console.log(user);

        const { uid, displayName } = user;
        dispatch(login(uid, displayName));
        dispatch(endLoadingAction());
      })
      .catch((e) => {
        console.log(e);
        dispatch(endLoadingAction());
        Swal.fire("Error", e.message, "error");
      });
  };
};

export const startRegisterWithEmailPasswordName = (email, password, name) => {
  return (dispatch) => {
    console.log(email, password, name);
    console.log(dispatch);
    firebase
      .auth()
      .createUserWithEmailAndPassword(email, password)
      .then(async ({ user }) => {
        await user.updateProfile({ displayName: name });

        const { uid, displayName } = user;
        dispatch(login(uid, displayName));
      })
      .catch((e) => {
        console.log(e);
        Swal.fire("Error", e.message, "error");
      });
  };
};

export const startGoogleLogin = () => {
  return (dispatch) => {
    firebase
      .auth()
      .signInWithPopup(googleAuthProvider)
      .then(({ user }) => {
        const { uid, displayName } = user;
        dispatch(login(uid, displayName));
      });
  };
};

export const login = (uid, displayName) => {
  return {
    type: types.login,
    payload: {
      uid,
      displayName,
    },
  };
};

export const startLogout = () => {
  return (dispatch) => {
    firebase
      .auth()
      .signOut()
      .then(() => {
        dispatch(logout());
      });
  };
};

export const logout = () => {
  return {
    type: types.logout,
  };
};
