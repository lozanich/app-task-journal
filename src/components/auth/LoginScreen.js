import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { Link } from "react-router-dom";
import { startLoginEmailPassword, startGoogleLogin } from "../../actions/auth";
import { useForm } from "../../hooks/useForm";

export const LoginScreen = () => {
  const dispatch = useDispatch();
  const state = useSelector((state) => state.ui);

  const { loading } = state;

  const { values, handleInputChange } = useForm({
    email: "loz@gmail.com",
    password: "123456",
  });

  const { email, password } = values;

  const handleLogin = (e) => {
    e.preventDefault();
    console.log(email, password);
    dispatch(startLoginEmailPassword(email, password));
  };

  const handleLoginGoogle = () => {
    console.log("login GOOGLE");

    dispatch(startGoogleLogin());
  };

  return (
    <>
      <h3 className="auth__title">Login</h3>

      <form onSubmit={handleLogin}>
        <input
          className="auth__input"
          type="text"
          placeholder="Email"
          name="email"
          autoComplete="off"
          value={email}
          onChange={handleInputChange}
        />
        <input
          className="auth__input"
          type="password"
          placeholder="Password"
          name="password"
          value={password}
          onChange={handleInputChange}
        />
        <button
          className="btn btn-primary btn-block"
          type="submit"
          disabled={loading}
        >
          Ingresar
        </button>
        <hr />
        <div className="auth__social-networks">
          <p>Login with social networks</p>
          <div className="google-btn" onClick={handleLoginGoogle}>
            <div className="google-icon-wrapper">
              <img
                className="google-icon"
                src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
                alt="google button"
              />
            </div>
            <p className="btn-text">
              <b>Sign in with google</b>
            </p>
          </div>
        </div>

        <Link className="link" to="/auth/register">
          Create new account
        </Link>
      </form>
    </>
  );
};
