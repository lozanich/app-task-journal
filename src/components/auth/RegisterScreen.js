import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { Link } from "react-router-dom";
import validator from "validator";
import { useForm } from "../../hooks/useForm";
import { setErrorAction, removeErrorAction } from "../../actions/ui";
import { startRegisterWithEmailPasswordName } from "../../actions/auth";

export const RegisterScreen = () => {
  const dispatch = useDispatch();
  const state = useSelector((state) => state.ui);

  const { msgError } = state;

  const { values, handleInputChange } = useForm({
    name: "Lozanich",
    email: "loz@gmail.com",
    password: "123456",
    password2: "123456",
  });

  const { name, email, password, password2 } = values;

  const handleRegister = (e) => {
    e.preventDefault();
    if (formValid()) {
      console.log("el formularios es valido");
      dispatch(startRegisterWithEmailPasswordName(email, password, name));
    }
    console.log(name, email, password, password2);
  };

  const formValid = () => {
    if (name.trim().length === 0) {
      dispatch(setErrorAction("Name is required"));
      return false;
    } else if (!validator.isEmail(email)) {
      dispatch(setErrorAction("Email is not valid"));
      return false;
    } else if (password !== password2 || password.length < 5) {
      dispatch(
        setErrorAction("Password should be equals and minimum 6 characters")
      );
      return false;
    }

    dispatch(removeErrorAction());

    return true;
  };

  return (
    <>
      <h3 className="auth__title">Register</h3>

      {msgError && <div className="auth__alert-error">{msgError}</div>}

      <form onSubmit={handleRegister}>
        <input
          className="auth__input"
          type="text"
          placeholder="Email"
          name="email"
          autoComplete="off"
          value={email}
          onChange={handleInputChange}
        />
        <input
          className="auth__input"
          type="text"
          placeholder="Name"
          name="name"
          autoComplete="off"
          value={name}
          onChange={handleInputChange}
        />
        <input
          className="auth__input"
          type="password"
          placeholder="Password"
          name="password"
          value={password}
          onChange={handleInputChange}
        />
        <input
          className="auth__input"
          type="password"
          placeholder="Confirm password"
          name="password2"
          value={password2}
          onChange={handleInputChange}
        />
        <button className="btn btn-primary btn-block mb-5" type="submit">
          Ingresar
        </button>

        <Link className="link" to="/auth/login">
          Already registered?
        </Link>
      </form>
    </>
  );
};
