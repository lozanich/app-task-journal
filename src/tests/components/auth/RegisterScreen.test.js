import "@testing-library/jest-dom";
import { mount } from "enzyme";
import configureStore from "redux-mock-store"; //ES6 modules
import thunk from "redux-thunk";
import { MemoryRouter } from "react-router-dom";
import { RegisterScreen } from '../../../components/auth/RegisterScreen'
import { Provider } from "react-redux";
import { types } from "../../../types/types";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
  auth: {},
  ui: { loading: false, msgError: null },
};

let store = mockStore(initState);
// store.dispatch = jest.fn()


const wrapper = mount(
    <Provider store={store}>
        <MemoryRouter>
            <RegisterScreen />
        </MemoryRouter>
    </Provider>
)

describe('Pruebas en el componente <RegisterScreen />', () => {
    beforeEach(() => {
        // store = mockStore(initState);
        jest.clearAllMocks()
    });

    test('Debe de hacer match con el snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    })

    test('Debe de hacer el dispatch de la accion respectiva', () => {
        const emailField = wrapper.find('input[name="email"]')
        emailField.simulate('change', {
            target: {
                value: '',
                name: 'email'
            }
        })

        wrapper.find('form').simulate('submit', {
            preventDefault(){}
        })

        const actions = store.getActions()

        expect(actions[0]).toEqual({
            type: types.uiSetError,
            payload: "Email is not valid"
        })

    })
    
    test('Debe de mostarr la caja de alerta con el error', () => {
        const initState = {
            auth: {},
            ui: { loading: false, msgError: "Email no es correcto" },
            };

        const store = mockStore(initState);

        const wrapper = mount(
            <Provider store={store}>
                <MemoryRouter>
                    <RegisterScreen />
                </MemoryRouter>
            </Provider>
        );
        expect(wrapper.find('.auth__alert-error').exists()).toBe(true)
        expect(wrapper.find('.auth__alert-error').text().trim()).toBe(initState.ui.msgError)

    })
    
    
    
})
