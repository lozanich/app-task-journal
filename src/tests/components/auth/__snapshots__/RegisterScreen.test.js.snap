// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <RegisterScreen /> Debe de hacer match con el snapshot 1`] = `
Array [
  <h3
    className="auth__title"
  >
    Register
  </h3>,
  <form
    className="animate__animated animate__fadeIn"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value="loz@gmail.com"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="Lozanich"
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value="123456"
    />
    <input
      className="auth__input"
      name="password2"
      onChange={[Function]}
      placeholder="Confirm password"
      type="password"
      value="123456"
    />
    <button
      className="btn btn-primary btn-block mb-5"
      type="submit"
    >
      Ingresar
    </button>
    <a
      className="link"
      href="/auth/login"
      onClick={[Function]}
    >
      Already registered?
    </a>
  </form>,
]
`;
